teps/0143-concise-parameters-results.md

```diff
@@ -0,0 +1,848 @@
+
+### Prior Art and Related Work
+
+There have been efforts in Tekton to make the API simple and concise.
+
+#### Mapping Workspaces
+
+Tekton auto-maps `Workspaces` from `Pipelines` to `PipelineTasks` when the names match, per [TEP-0108](0108-mapping-workspaces.md) to reduce the verbosity in mapping `Workspaces`.
+
+#### Propagated Workspaces and Parameters
+
+Tekton propagates `Workspaces` and `Parameters` in embedded specifications, per [TEP-0111](0111-propagating-workspaces.md) and [TEP-0107](0107-propagating-parameters.md) to reduce the verbosity of repeatedly defining `Workspaces` and `Parameters`.
+
+#### Github Actions
+
+Github Actions uses maps for `Inputs` and `Outputs`. Take the [Cloud Functions Deploy][cfd] for example:
+
+```yaml
+name: Cloud Functions Deploy
+description: Use this action to deploy your function or update an existing Cloud Function.
+inputs:
+  name:
+    description: Name of the Cloud Function.
+    required: true
+  description:
+    description: Description for the Cloud Function.
+    required: false
+  runtime:
+    description: Runtime to use for the function.
+    required: true
+outputs:
+  url:
+    description: The URL of your Cloud Function. Only available with HTTP Trigger.
+  id:
+    description: Full resource name of the Cloud Function, of the format 'projects/p/locations/l/functions/f'.
+  status:
+    description: Status of the Cloud Function deployment.
+  version:
+    description: Version of the Cloud Function deployment.
+  runtime:
+    description: Runtime of the Cloud Function deployment.
+runs:
+  using: 'node16'
+  main: 'dist/index.js'
+```
+
+## Proposal
+
+We propose using maps for `Params` and `Results` instead of name and value pairs.
+
+### Syntax
+
+#### Parameters
+
+At authoring time, users declare string, array and object `Parameters` for `Tasks` and `Pipelines`. It is tedious to specify these fields because they are defined as sub-objects. We propose using maps instead of sub-objects to declare `Parameters`. The new syntax of `Params` will go under a new api field `Inputs.Params`. This is because we cannot modify the existing `Params` field because of our [Go compatibility policy][cp].
+
+```yaml
+# before
+params:
+  - name: hello
+    type: string
+  - name: environments
+    type: array
+  - name: git
+    type: object
+    properties:
+      url: {}
+      commit: {}
+
+# after
+inputs:
+  params:
+    hello:
+      type: string
+    environments:
+      type: array
+    git:
+      type: object
+      properties:
+        url: {}
+        commit: {}
+```
+
+At runtime, users configure execution by passing in `Parameters`. We propose using maps instead of sub-objects to declare `Parameters` at runtime.
+
+```yaml
+# before
+params:
+  - name: hello
+    value: world
+  - name: environments
+    value:
+      - 'staging'
+      - 'qa'
+      - 'prod'
+  - name: git
+    value:
+      url: abc.com
+      commit: sha123
+
+# after
+inputs:
+  params:
+    hello: world
```

Comment:
This is unsafe. I assume it will be decoded/unmarshal to:
```
map[string]interface{}
 ```
 ? Yes?

Reply:
What happens is that we might need to have some reserved keywords for the string `key`.
What happens if someone mistakenly indents [serviceAccountName](https://tekton.dev/docs/pipelines/pipelineruns/#specifying-custom-serviceaccount-credentials) at the same level? Or some other top elements. 

The present API won't let you create these.

Reply:
but the all these `key: value`, fields defined under `params` right?
`serviceAccountName` comes under `params`?

Reply:
`serviceAccountName` doesn't come under `params`. It's on a different level but here we don't know whether key `serviceAccountName` is some params or spec.serviceAccountName . So we can't validate.

Reply:
On a side note, we shouldn't think of this from trying to solve validation via Go code. Is it easier to validations in 1.26+ k8s using CEL Validating admission policies?

Reply:
> This is unsafe. I assume it will be decoded/unmarshal to:
> 
> ```
> map[string]interface{}
> ```
> 
> ? Isn't it so?

We will still use this [function](https://github.com/tektoncd/pipeline/blob/f8c3a3a7ba7afbad1b9bc97c90f480ea89eac17b/pkg/apis/pipeline/v1/param_types.go#L434) to unmarshal it, it will be map[string]ParamValue

Reply:
@khrm @Yongxuanzhang I'm not 100% I got it right, this is my understanding of the issue.
Current `TaskRun` spec:

```
spec:
  params:
  - name: example
    value: foo
  serviceAccountName: default
```

New `TaskRun` spec:

```
spec:
  input:
    params:
      example: foo
  serviceAccountName: default
```

With the new spec, if `serviceAccountName` was not indented properly,  you could have:

```
spec:
  input:
    params:
      example: foo
      serviceAccountName: default
```

Which would not fail validation if the `Task` in question had a parameter called `serviceAccountName`.
This could not happen with the current API. This could be prevented in the new syntax by forbidding `serviceAccountName` as a parameter name.